pkg/client/cache/expiration_cache.go

```diff
@@ -0,0 +1,178 @@
+// so unittests don't rely on the system clock.
+type ExpirationPolicy interface {
+	IsExpired(obj interface{}) bool
+}
+
+// TTLPolicy implements a ttl based ExpirationPolicy.
+type TTLPolicy struct {
+	//	 >0: Expire entries with an age > ttl
+	//	<=0: Don't expire any entry
+	ttl time.Duration
+}
+
+// IsExpired returns true if the given object is older than the ttl, or it can't
+// determine its age.
+func (p *TTLPolicy) IsExpired(obj interface{}) bool {
+	if timeStampedEntry, ok := obj.(*TimeStampedEntry); !ok {
+		return true
+	} else {
+		return p.ttl <= 0 || time.Since(timeStampedEntry.timeStamp) > p.ttl
+	}
+}
+
+// TimeStampedEntry is the only type allowed in a ExpirationCache.
+type TimeStampedEntry struct {
+	obj       interface{}
+	timeStamp time.Time
+}
+
+// getTimeStampedEntry returnes the TimeStampedEntry stored under the given key.
+func (c *ExpirationCache) getTimeStampedEntry(key string) (*TimeStampedEntry, bool) {
+	item, exists := c.cacheStorage.Get(key)
+	// TODO: Check the cast instead
+	if exists {
+		return item.(*TimeStampedEntry), exists
+	}
+	return nil, false
+}
+
+// getOrExpire retrieves the object from the TimeStampedEntry iff it hasn't
+// already expired. It kicks-off a go routine to delete expired objects from
+// the store and sets exists=false.
+func (c *ExpirationCache) getOrExpire(key string) (interface{}, bool) {
+	timeStampedItem, exists := c.getTimeStampedEntry(key)
+	if !exists {
+		return nil, false
+	}
+	if c.expirationPolicy.IsExpired(timeStampedItem) {
+		go func() { c.cacheStorage.Delete(key) }()
+		return nil, false
+	}
+	return timeStampedItem.obj, true
+}
+
+// GetByKey returns the item stored under the key, or sets exists=false.
+func (c *ExpirationCache) GetByKey(key string) (interface{}, bool, error) {
+	obj, exists := c.getOrExpire(key)
+	return obj, exists, nil
+}
+
+// Get returns unexpired items. It purges the cache of expired items in the
+// process.
+func (c *ExpirationCache) Get(obj interface{}) (interface{}, bool, error) {
+	key, err := c.keyFunc(obj)
+	if err != nil {
+		return nil, false, err
+	}
+	obj, exists := c.getOrExpire(key)
+	return obj, exists, nil
+}
+
+// List retrieves a list of unexpired items. It purges the cache of expired
+// items in the process.
+func (c *ExpirationCache) List() []interface{} {
+	items := c.cacheStorage.List()
+
+	list := make([]interface{}, 0, len(items))
+	for _, item := range items {
+		obj := item.(*TimeStampedEntry).obj
+		if key, err := c.keyFunc(obj); err != nil {
+			list = append(list, obj)
+		} else if obj, exists := c.getOrExpire(key); exists {
+			list = append(list, obj)
+		}
+	}
+	return list
+}
+
+// Add inserts a timestamped item into the cache, overwriting entries that might
+// exist under the same key.
+func (c *ExpirationCache) Add(obj interface{}) error {
+	key, err := c.keyFunc(obj)
+	if err != nil {
+		return fmt.Errorf("couldn't create key for object: %v", err)
+	}
```

Comment:
Suggestion: check type of obj; if it happens to be a timestampedEntry, don't wrap it in another timestamped entry. Has nice side effect of letting people provide the time themselves if they wish. Document this.

Also fix the error type returned here.